src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,161 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final Object[] DEFAULTCAPACITY_EMPTY_ELEMENTDATA = {};
+    private static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE - 8;
+    private static final int startSize = 0;
+    private Object[] elementData;
+    private int size;
+
+    public ArrayList() {
+        size = startSize;
+        this.elementData = DEFAULTCAPACITY_EMPTY_ELEMENTDATA;
+    }
+
+    private void ensureCapacityInternal(int minCapacity) {
+        ensureExplicitCapacity(calculateCapacity(elementData, minCapacity));
+    }
+
+    private static int calculateCapacity(Object[] elementData, int minCapacity) {
+        if (elementData == DEFAULTCAPACITY_EMPTY_ELEMENTDATA) {
+            return Math.max(DEFAULT_CAPACITY, minCapacity);
+        }
+        return minCapacity;
+    }
+
+    private void ensureExplicitCapacity(int minCapacity) {
+        if (minCapacity - elementData.length > 0) {
+            grow(minCapacity);
+        }
+    }
+
+    private void arrayListIndexOutOfBounds(int index, boolean mode) {
+        if (mode) {
+            if (index > size || index < 0) {
+                throw new ArrayListIndexOutOfBoundsException("Element not found in the list");
+            }
+        } else {
+            if (index >= size || index < 0) {
+                throw new ArrayListIndexOutOfBoundsException("Element not found in the list");
+            }
+        }
+    }
+
+    private void grow(int minCapacity) {
+        int oldCapacity = elementData.length;
+        int newCapacity = oldCapacity + (oldCapacity >> 1);
```

Comment:
The growth factor `1.5` is used here. It would be better to define this as a constant to improve code readability and maintainability.